### ISSUE TITLE
Windows 7 32 Bits Error

### ISSUE THREAD

**Issue by FrankSanabria** [Ylianst/MeshCentral#1426]
Meshcentral version: 0.5.46
SO Guest: Windows 7 32 Bits Lastest Updates

Hi, We have this error with windows 7 32 bits installation, using GUI and Console.

```
C:\Users\sugeek\Downloads>meshagent32-Externos.exe -fullinstall
MeshCentral2 Agent
...Checking for previous installation [NONE]
...Installing service [ERROR] fs.mkdirSync(): Unable to create dir: undefined\me
sh
C:\Users\sugeek\Downloads>

```
Thanks for your work

**Comment by krayon007** [comment]
if you open a command prompt, what does this return:
echo %ProgramFiles%

Looking at the output you posted, it looks like the environment variable for the ProgramFiles folder was undefined.

I'm guessing on your platform, the environment variable name is not in english... I'll see if I can find how to fetch it correctly, if you can verify my hypothesis is correct.

**Comment by ryanblenis** [comment]
May or may not be at all related, but a prior version (I believe 0.5.37 IIRC) for some reason, on some Win10 machines, did install under a C:\Program Files\mesh\Mesh Agent\... directory instead of the usual C:\Program Files\Mesh Agent\... Not sure how to check but one of my test computers still has both directories.

**Comment by krayon007** [comment]
@ryanblenis, your issue is because I changed the installer to use JS to install the agent. Before, the agent binary on windows would manually install itself into Mesh Agent. But I had to remove it, becuase it didn't work on UTF8 paths. So now that it's switched to using JS to do the install, the same code works across all platforms, which is why you can now do --fullinstall on any platform now, not just windows...

So on windows, the 'mesh' folder is equivalent to 'mesh_services' on the other platforms... In retrospect, I probably should've made it 'mesh_services' across the board, but Ylian already had some other things using the 'mesh' folder, so I just adopted that for windows..... As far as the Windows 7 32 bit thing, I'm working on trying to reproduce it. I can't reproduce it on my Windows 7 64 bit build, so I'm going to build a new test setup. I'm trying to figure out if it's a win 7 32 bit issue, a Service Pack issue, or a locale issue. According to the msdn docs, the environment variable I used should be defined in Windows 7 32 bit, which is why I suspect the variable name might be subject to locale as well...

**Comment by krayon007** [comment]
In the meantime, on windows, if you are installing from the command line, add the: --installPath switch, and specify the path. Like:

meshagent32-Externos.exe -fullinstall --installPath="C:\\My\\Folder"

Currently when doing backslashes, you have to escape them like in the example above.
Also, currently, specifying alternative install paths is only supported on Windows, becuase on other platforms I ran into issues with file system permissions, so I left it as a windows only feature for now.

**Comment by krayon007** [comment]
Just as a note, the backslashes need to be doubled up... They weren't viewing correctly on the previous post, so I had to escape it again, so it would render on the html page correctly...

**Comment by FrankSanabria** [comment]
Thanks, With --installPath Option now work. 

The Variable %ProgramFiles% is equal to C:\Program Files

**Comment by krayon007** [comment]
> Thanks, With --installPath Option now work.
> 
> The Variable %ProgramFiles% is equal to C:\Program Files

Can you try the following command from the command prompt, wherever your agent binary is:
MeshAgent -exec "console.log(require('os').arch(),process.env['ProgramFiles']);process.exit();"

I'm trying to figure out how the installer got 'undefined' as the ProgramFiles location.

**Comment by krayon007** [comment]
> Meshcentral version: 0.5.46
> SO Guest: Windows 7 32 Bits Lastest Updates
> 
> Hi, We have this error with windows 7 32 bits installation, using GUI and Console.
> 
> ```
> C:\Users\sugeek\Downloads>meshagent32-Externos.exe -fullinstall
> MeshCentral2 Agent
> ...Checking for previous installation [NONE]
> ...Installing service [ERROR] fs.mkdirSync(): Unable to create dir: undefined\me
> sh
> C:\Users\sugeek\Downloads>
> ```
> 
> Thanks for your work

Also, what locale and version is your Windows 7?

**Comment by krayon007** [comment]
Ok, I setup Brazillian Portuguese Windows 7 SP1 32-bit, and was able to reproduce the issue...
The root cause appears to be a bug in the Win32 API.. I'll have to check if this affects all Windows 7 32 bit, or just locale variants... 

But the issue is that IsWOW64Process() is not supposed to exist on 32 bit windows... But it does exist on this version of windows... When I dynamically load it, and try to test the agent process.... The MSDN documentation says that on 32-bit windows this returns FALSE. It is only supposed to return TRUE if the process is a 32 bit process running on 64-bit Windows...

I verified that my Brazillian Portuguese Windows 7, is in fact 32 bit, but IsWow64Process() is returning TRUE. So the agent, thinks it's a 32 bit service being installed on 64 bit windows, so it is trying to install to %ProgramFiles(x86)% which isn't defined (as it's not supposed to be).

As a work-around, I'll have it use %ProgramFiles% if %ProgramFiles(x86)% isn't defined...
I'll also look into if this is a broader Windows 7 issue or a locale specific issue...

**Comment by krayon007** [comment]
Nevermind, I found the actual issue... There are two BOOL values in IsWow64Process(), and I was looking at the wrong one, lol...

**Comment by krayon007** [comment]
I fixed this in the agent repo. It will be in the next agent update.

**Comment by Ylianst** [comment]
Published MeshCentral v0.5.51 with Bryan's all new MeshAgents. Should fix this issue.
